### PR TITLE
OneFlow Serving

### PR DESCRIPTION
**Pull request by leaves-zwx**
## OneFlow Serving Roadmap

- [x] save model
  - [x] saved_model_proto
  - [x] saved_model_builder
- [x] load model
  - [x] InferenceSession
  - [x] refine op graph by job signature
  - [ ] ~~support job graph (@Ldpe2G ) (2020-11-03)~~ (暂不支持)
  - [ ] user specified env (@leaves-zwx ) (2020-11-27)
- [ ] oneflow server
  - [x] model server (@leaves-zwx )
  - [x] async grpc server (@leaves-zwx )
  - [ ] more grpc service (GetModelDesc etc.) (@leaves-zwx )
  - [ ] support multi model serve
  - [ ] async http server (@Ldpe2G ) (2020-11-29)
  - [ ] batching (@Ldpe2G ) (2020-11-29)
  - [ ] version policy (@leaves-zwx )
  - [ ] hot update (1. config file watch, 2. reload protocol/service) (@leaves-zwx )
  - [ ] other: bulk processing (streaming); credential; improve performance on cpu and gpu

### SavedModelBuilder

```
    saved_model_builder_v2 = flow.SavedModelBuilderV2("./saved_models_v2")
    saved_model_builder_v2.ModelName("alexnet").Version(1).Job(alexnet_eval_job).Input(
        "image", "CropMirrorNormalize_14/out_0"
    ).Input("label", "Squeeze_12/out_0").Output(
        "score", "softmax_loss/out_0"
    ).Complete().Save()
```

待补充

### SimpleSession (InferenceSession)

```
sess = flow.SimpleSession()
sess.set_checkpoint_path(checkpoint_path)
for job_name, signature in saved_model_proto.signatures_v2.items():
    sess.setup_job_signature(job_name, signature)
for job_name, net in saved_model_proto.graphs.items():
    with sess.open(job_name) as sess:
        sess.compile(net.op)
sess.launch()
for i, (image, label) in enumerate(zip(image_list, label_list)):
    outputs = sess.run(alexnet_infer.__name__, image=image, label=label)
```

待补充

### model server

待补充

### user protocol

待补充

## OneFlow Serving Design

待补充

## 参考

[tensorflow serving simple example](https://www.tensorflow.org/tfx/tutorials/serving/rest_simple)
[pytorch saving and loading models example](https://pytorch.org/tutorials/beginner/saving_loading_models.html)
之江 TS_Serving

### CODE REVIEW

File: oneflow/core/serving/predict.proto [hunk 1]
```diff
@@ -0,0 +1,18 @@
+syntax = "proto2";
+package oneflow.serving;
+
+import "oneflow/core/serving/model_spec.proto";
+import "oneflow/core/serving/tensor.proto";
+
+message PredictRequest {
+  required ModelSpec model_spec = 1;
+  map<string, TensorProto> inputs = 2;
```
Comment (Ldpe2G) [inline oneflow/core/serving/predict.proto:9]: 这样子设计感觉对于client的每一次request，对于网络的没一个输入只能上传一个 tensor，如果用户一次想请求预测一个batch的图片，这样子是不能支持的？

Comment (leaves-zwx) [reply oneflow/core/serving/predict.proto:9]: 这里是个 map，是可以传入多个 input 的，比如同时传入 image 和 label。预测一个 batch 的图片的话，不就是放在一个 tensor 里面吗？

Comment (Ldpe2G) [reply oneflow/core/serving/predict.proto:9]: 我懂了，不过这样子就只能支持client端的一个batch的图片都是同等大小的？

```protobuf
message TensorProto {
  required ShapeProto shape = 1;
  required DataType data_type = 2;
  oneof data {
    BytesList bytes_list = 11;
    FloatList float_list = 12;
    DoubleList double_list = 13;
    Int32List int32_list = 14;
    Int64List int64_list = 15;
  }
}
```

Comment (leaves-zwx) [reply oneflow/core/serving/predict.proto:9]: 是的，如果是以 batch 为单位还是建议图片都是同等大小，否则的话，还是一张一张的推理吧。

File: oneflow/python/serving/simple_session.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+
+        return output_lbn, output_job_name
+
+    def input_info(self, input_name):
+        self._check_status(self.SessionStatus.OPEN, self.SessionStatus.RUNNING)
+        input_lbn, input_job_name = self._query_input_lbn(input_name)
+        if input_job_name is None or input_lbn is None:
+            raise ValueError('can not find input "{}"'.format(input_name))
+
+        input_shape = c_api_util.JobBuildAndInferCtx_GetStaticShape(
```
Comment (Ldpe2G) [inline oneflow/python/serving/simple_session.py:244]: 发现一个问题，`JobBuildAndInferCtx_GetStaticShape` 最底层是调用的  `JobBuildAndInferCtx::GetStaticShape`

```C++
Maybe<Shape> JobBuildAndInferCtx::GetStaticShape(const std::string& lbn) const {
  JUST(CheckLbnValidAndExist(lbn));
  if (this->job_ == nullptr) { printf("job is nullptr\n"); }
  return lbi2logical_blob_desc_.at(GenLogicalBlobId(lbn))->shape();
}
```

lbi2logical_blob_desc_ 里面的内容只会在 addandinferop 的时候被修改，后面即使再运行改图 pass 的时候也没再更新里面的 blob_desc，所以从中获取到的输入 blob 的 shape 任然是 saved_model 里面原始的 shape，但是实际底层的 input_blob 是已经改为了和 input_signature 的 shape 了。

我是尝试给 grpc_demo 添加设置 batch_size 的时候发现的问题。

Comment (leaves-zwx) [reply oneflow/python/serving/simple_session.py:244]: 我对这里的设想是在 load model 实际执行之前，需要根据 batch_size 把 job_conf 里面的 signature 中的 input_def 的形状修改掉。 

```
message JobConfigProto {
  ...
  optional JobSignatureDef signature = 4000;
}

Comment (Ldpe2G) [reply oneflow/python/serving/simple_session.py:244]: 我就是改这里的，这里改了，最终 run 的时候确实是设置的 batch 大小。
但是get_input_info 返回的还是没改之前的。
简单来说上面从 c_api_util 获取的 input shape 是 saved_model 里面保存的 graph 的 input shape

```proto
message SavedModel {
  // .....
  map<string, DLNetConf> graphs = 4;
  map<string, JobSignatureDef> signatures_v2 = 6;
}
```

Comment (leaves-zwx) [reply oneflow/python/serving/simple_session.py:244]: input_info 在 sess.launch 之前取的还是之后取的？

Comment (Ldpe2G) [reply oneflow/python/serving/simple_session.py:244]: sess.launch 之前调用不了，实现要求 session 的状态是  RUNNING

Comment (Ldpe2G) [reply oneflow/python/serving/simple_session.py:244]: 是这里，c_api_util.JobBuildAndInferCtx_GetStaticShape 返回的是  compile_ctx.CurJobAddOp 时候的 shape，而之后
c_api_util.CurJobBuildAndInferCtx_Complete 所修改的 input shape 没有更新，所以获取了旧的

```python3
    def compile(self, op_list):
        self._check_status(self.SessionStatus.OPEN)
        for op_conf in op_list:
            compile_ctx.CurJobAddOp(op_conf)

        c_api_util.CurJobBuildAndInferCtx_Complete()
```

Comment (leaves-zwx) [reply oneflow/python/serving/simple_session.py:244]: 嗯，我明白这里的问题了，JobBuildAndInferCtx_GetStaticShape 这个 api 取的是 job 中的数据。新的 batch_size 在被设置后，oneflow 对象构建的时候会重新构造新的 op_graph (根据新的 InterfaceBlobConf)，这里所有 op 的 InferBlobDescs 都会被重新调用，所以能保证实际 run 的时候 shape 都是正常的。但 job_build_and_infer_ctx 这个里面 shape 还是旧的，所以通过 JobBuildAndInferCtx_GetStaticShape 接口取到的 shape 是旧的。我想一下这个地方有办法解决不。

Comment (Ldpe2G) [reply oneflow/python/serving/simple_session.py:244]: 是这个问题

Comment (daquexian) [reply oneflow/python/serving/simple_session.py:244]: （乱入）https://github.com/Oneflow-Inc/oneflow/commit/26f04c13c39dc2d24d44d65c73ba58aa0e06ccb1 这个 api 会有用吗，拿到的是 infer 之后的所有 OpAttribute

Comment (Ldpe2G) [reply oneflow/python/serving/simple_session.py:244]: 好的，看下先

**Comment by leaves-zwx** [comment]
## 20201125 oneflow 之江进度会议记录

12月中旬，之江需要进行首次验收，目前支持的 saved model 不够丰富，已支持的模型: 

- resnet50
- alexnet
- inception V3 (结果错误)
- vgg16

cpu 版本目前缺少 openvino 的支持，暂时先只支持 gpu 版本的 serving

未来需要进一步支持的推理模型（在 oneflow-benchmark 中）

- mobilenet
- bert
- wide_and_deep

model server 未来的改进方向，目前之江对 model server 的需求不是很大，主要依赖 saved model 和 simple session。提高完善 saved_model 和 simple session 功能的任务的优先级，考虑 saved_model 和 simple session 在支持其他的模型的时候有什么局限性。在改进 model server 的过程中，考虑怎么改进 saved_model 和 simple_session 以增强其兼容性和灵活性。
